File: engine/src/main/java/top/skyeyefast/mchjong/engine/HandBonuses.java
package top.skyeyefast.mchjong.engine;

import java.util.List;

/** The same physical red, indicator and north bonuses for settlement and hypothetical wins. */
final class HandBonuses {
    private HandBonuses() {}
    static int[] indicators(List<Integer> indicators, boolean sanma) {
        int[] counts = new int[34];
        for (int tile : indicators) if (tile >= 0) counts[Tile.doraAfter(Tile.kind(tile), sanma)]++;
        return counts;
    }
    static int tile(int tile, int[] dora) { return dora[Tile.kind(tile)] + (Tile.red(tile) ? 1 : 0); }
    static int count(List<Integer> hand, List<Meld> melds, List<Integer> norths, int winning, int[] dora) {
        int count = norths.size();
        if (winning >= 0 && hand.size() + melds.size() * 3 == 13) count += tile(winning, dora);
        for (int tile : hand) count += tile(tile, dora);
        for (var meld : melds) for (int tile : meld.tiles()) count += tile(tile, dora);
        for (int tile : norths) count += tile(tile, dora);
        return count;
    }
}

File: engine/src/main/kotlin/top/skyeyefast/mchjong/engine/HandBonuses.kt
package top.skyeyefast.mchjong.engine

/** The same physical red, indicator and north bonuses for settlement and hypothetical wins. */
internal object HandBonuses {
    @JvmStatic
    fun indicators(indicators: List<Int>, sanma: Boolean): IntArray {
        val counts = IntArray(34)
        for (tile in indicators) if (tile >= 0) counts[Tile.doraAfter(Tile.kind(tile), sanma)]++
        return counts
    }

    fun tile(tile: Int, dora: IntArray): Int = dora[Tile.kind(tile)] + if (Tile.red(tile)) 1 else 0

    @JvmStatic
    fun count(hand: List<Int>, melds: List<Meld>, norths: List<Int>, winning: Int, dora: IntArray): Int {
        var count = norths.size
        if (winning >= 0 && hand.size + melds.size * 3 == 13) count += tile(winning, dora)
        for (tile in hand) count += tile(tile, dora)
        for (meld in melds) for (tile in meld.tiles()) count += tile(tile, dora)
        for (tile in norths) count += tile(tile, dora)
        return count
    }
}
